pennylane/qinfo/transforms.py

```diff
@@ -396,3 +396,176 @@ def wrapper(*args, **kwargs):
         return _compute_cfim(p, j)
 
     return wrapper
+
+
+def fidelity(qnode0, qnode1, wires0, wires1):
+    r"""Compute the fidelity for two :class:`.QNode` returning a :func:`~.state` (a state can be a state vector
+    or a density matrix, depending on the device) acting on quantum systems with the same size.
+
+    The fidelity for two mixed states given by density matrices :math:`\rho` and :math:`\sigma`
+    is defined as
+
+    .. math::
+        F( \rho , \sigma ) = \text{Tr}( \sqrt{\sqrt{\rho} \sigma \sqrt{\rho}})^2
+
+    If one of the states is pure, say :math:`\rho=\ket{\psi}\bra{\psi}`, then the expression
+    for fidelity simplifies to
+
+    .. math::
+        F( \ket{\psi} , \sigma ) = \bra{\psi} \sigma \ket{\psi}
+
+    Finally, if both states are pure, :math:`\sigma=\ket{\phi}\bra{\phi}`, then the
+    fidelity is simply
+
+    .. math::
+        F( \ket{\psi} , \ket{\phi}) = \left|\braket{\psi, \phi}\right|^2
+
+    .. note::
+        The second state is coerced to the type and dtype of the first state. The fidelity is returned in the type
+        of the interface of the first state.
+
+    Args:
+        state0 (QNode): A :class:`.QNode` returning a :func:`~.state`.
+        state1 (QNode): A :class:`.QNode` returning a :func:`~.state`.
+        wires0 (Sequence[int]): the wires of the first subsystem
+        wires1 (Sequence[int]): the wires of the second subsystem
+
+    Returns:
+        func: A function that returns the fidelity between the states outputted by the QNodes.
+
+    **Example**
+
+    First define two QNodes with potentially different signatures, let's consider a first circuit with two parameters
+    and a second circuit with only one parameter. In the `qml.qinfo.fidelity` transform it is then necessary to give
+    two tuple, each containing the args and kwargs of their respective circuit, e.g. `all_args0` = (0.1, 0.3) and
+    `all_args1` = (0.2) in the following.
```

Comment:
Suggestion for wording. In addition, it could be nice to show a circuit with a kwarg too.
```suggestion
    First, let's consider two QNodes with potentially different signatures: a circuit with two parameters
    and another circuit with a single parameter. The output of the `qml.qinfo.fidelity` transform then requires
    two tuples to be passed as arguments, each containing the args and kwargs of their respective circuit, e.g. `all_args0 = (0.1, 0.3)` and
    `all_args1 = (0.2)` in the following case:
```